### PR TITLE
#1442 Added retry logic for get threads list and thread info

### LINKED ISSUE #1442: request timed out when loading inbox
image

We do a lot of requests at once when loading inbox, and it may sometimes time out.

When it times out, we seem to fail the whole loading and then we need to make all of the requests again.

I think it would be better if only the email messages that timed out had the failure showing on them, with a retry button.

Then retrying one of them would retry all of the failed ones. That's assuming that the "list of messages" API call succeeded, and it was some individual messages that failed. The initial list of messages (returned successfully) could result in "empty list" of messages already pre-rendered, and then individual message details could be rendered into that list as they become available. That way, the loading will become more interactive.

This way, at least, most of the inbox will be loaded when timeout happens on one or two messages.

After that, we can investigate in more detail how to make this happen less often. Maybe some automatic retry approach.

### PR DESCRIPTION
**Pull request by ioanmo226**
This PR added retry logic for Task level and added this logic in get threads list and get thread info calls.
Default retry count is `3` and can be changed by passing `maxRetryCount` parameter.

close #1442 // if this PR closes an issue

----------------------------------

**Tests** _(delete all except exactly one)_:
- Difficult to test (not sure what & how to test)

--------------------------------

### To be filled by reviewers

I have reviewed that this PR... _(tick whichever items you personally focused on during this review)_:
- [x] addresses the issue it closes (if any)
- [x] code is readable and understandable
- [x] is accompanied with tests, or tests are not needed
- [x] is free of vulnerabilities
- [x] is documented clearly and usefully, or doesn't need documentation

### CODE REVIEW

**Comment by mobilestar1** [comment]
For above case, when 19 threads are loaded and last one fails, app retries to fetch only that last message(not whole 20 threads).
App only tries to re-fetch failed threads (up to 3 retries).

image

**Comment by ioanmo226** [comment]
For above case, when 19 threads are loaded and last one fails, app retries to fetch only that last message(not whole 20 threads).
App only tries to re-fetch failed threads (up to 3 retries).

image

**Comment by tomholub** [comment]
That's better. I'll take another look.

File: FlowCrypt/Functionality/Error Handling/Task+Retry.swift [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+//
+//  Created by Ioan Moldovan on 3/30/22
+//  Copyright © 2017-present FlowCrypt a. s. All rights reserved.
+//
+
+extension Task where Failure == Error {
+    @discardableResult
+    static func retrying(
+        priority: TaskPriority? = nil,
+        maxRetryCount: Int = 3,
```
Comment (tomholub) [inline FlowCrypt/Functionality/Error Handling/Task+Retry.swift:13]: Let's do a default of 2. So three API calls total maximum. Current setting with 3 may cause up to 4 API calls which is a bit too much.

File: FlowCrypt/Functionality/Error Handling/Task+Retry.swift [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        Task(priority: priority) {
+            for _ in 0 ..< maxRetryCount {
+                do {
+                    return try await operation()
+                } catch {
+                    let oneSecond = TimeInterval(1_000_000_000)
+                    let delay = UInt64(oneSecond * retryDelay)
+                    try await Task<Never, Never>.sleep(nanoseconds: delay)
+
+                    continue
```
Comment (tomholub) [inline FlowCrypt/Functionality/Error Handling/Task+Retry.swift:26]: If you are at the end of a loop block, I don't think `continue` would have any effect.

**Review by tomholub — CHANGES_REQUESTED** [review]
This looks good - please see comments below

File: FlowCrypt/Functionality/Error Handling/Task+Retry.swift [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+//  Created by Ioan Moldovan on 3/30/22
+//  Copyright © 2017-present FlowCrypt a. s. All rights reserved.
+//
+
+extension Task where Failure == Error {
+    @discardableResult
+    static func retrying(
+        priority: TaskPriority? = nil,
+        maxRetryCount: Int = 3,
+        retryDelay: TimeInterval = 1,
```
[the remark this answers is not in the document]
Comment (ioanmo226) [reply FlowCrypt/Functionality/Error Handling/Task+Retry.swift:14]: I agree

File: FlowCrypt/Functionality/Error Handling/Task+Retry.swift (hunk 1 above)
Comment (ioanmo226) [reply FlowCrypt/Functionality/Error Handling/Task+Retry.swift:13]: OK

File: FlowCrypt/Functionality/Error Handling/Task+Retry.swift (hunk 2 above)
Comment (ioanmo226) [reply FlowCrypt/Functionality/Error Handling/Task+Retry.swift:26]: Right. I added `continue` to make intent more clear.
Do you want me to remove that?

Comment (tomholub) [reply FlowCrypt/Functionality/Error Handling/Task+Retry.swift:26]: yes - please remove

Comment (ioanmo226) [reply FlowCrypt/Functionality/Error Handling/Task+Retry.swift]: OK

**Review by tomholub — APPROVED** [review]

**Pull request merged**